app/src/main/java/unipd/se18/ocrcamera/PhotoTester.java

```diff
@@ -0,0 +1,449 @@
+                e.printStackTrace();
+            }
+
+        }
+
+        //wait for all tests to complete
+        try {
+            countDownLatch.await();
+        } catch (InterruptedException e) {
+            e.printStackTrace();
+        }
+
+        long ended = java.lang.System.currentTimeMillis();
+        Log.i(TAG,"testAndReport ended (" + totalTestElements + " pics tested in " + (ended - started) + " ms)");
+
+        String report = jsonReport.toString();
+
+        //write report to file
+        try {
+            writeReportToExternalStorage(report, dirPath, "report.txt");
+        } catch (IOException e) {
+            Log.e(TAG, "Error writing report to file.");
+            e.printStackTrace();
+        }
+
+        return report;
+    }
+
+    public TestElement[] getTestElements() {
+        return testElements.toArray(new TestElement[0]);
+    }
+
+
+    /**
+     * Compare the list of ingredients extracted by OCR and the correct list of ingredients
+     * @param correct correct list of ingredients loaded from file
+     * @param extracted list of ingredients extracted by the OCR
+     * @return confidence percentage based on number of matched words, their similarity and order
+     * @author Francesco Pham credit to Stefano Romanello for Levenshtein library suggestion
+     */
+    private float ingredientsTextComparison(String correct, String extracted){
+
+        extracted = extracted.toLowerCase(); //ignoring case
+
+        //split words
+        String[] extractedWords = extracted.trim().split("[ ,-:./\\n\\r]+");
+        String[] correctWords = correct.trim().split("[ ,-:./\\n\\r]+");
+
+        Log.i(TAG, "ingredientsTextComparison -> Start of comparing");
+        Log.i(TAG, "ingredientsTextComparison -> correctWords.length == " + correctWords.length + ", extractedWords.length == " + extractedWords.length);
+
+        float points = 0;
+        int maxPoints = 0;
+        int posLastWordFound = 0;
+        int consecutiveNotFound = 0;
+
+
+        WeightedLevenshtein levenshtein = new WeightedLevenshtein(
+                new CharacterSubstitutionInterface() {
+                    public double cost(char c1, char c2) {
+
+                        // The cost for substituting 't' and 'r' is considered
+                        // smaller as these 2 are located next to each other
+                        // on a keyboard
+                        if (c1 == 't' && c2 == 'r') {
+                            return 0.5;
+                        }
+                        else if (c1 == 'q' && c2 == 'o') {
+                            return 0.5;
+                        }
+                        else if (c1 == 'I' && c2 == 'l') {
+                            return 0.5;
+                        }
+
+                        // For most cases, the cost of substituting 2 characters
+                        // is 1.0
+                        return 1.0;
+                    }
```

Comment:
It doesn't make sense to replace them with final variables and define them elsewhere in the code because that is where I am overriding the cost function for the Levenshtein algorithm.